File: Jewels And Stones.cc
class Solution {
public:
    int hash(char c) {
        if (c >= 'A' and c <= 'Z')  return 26 + c - 'A';
	    else return c - 'a';
    }
    
    int numJewelsInStones(string J, string S) {
        bool HTable[52] = {false};

        for (char c: J) {
            int key = hash(c);
            HTable[key] = true;
        }

        int ans = 0;
        for (char c: S) {
            int key = hash(c);
            if (HTable[key]) ans ++;
        }

        return ans;  
    }
};

File: Jewels And Stones.java
//using own Hashing
class Solution 
{
    public int hash(char key)
    {
        if (key >= 'A' && key <= 'Z')  return 26 + key - 'A';
	      else return key - 'a';
    }

    public int numJewelsInStones(String J, String S) 
    {
        int[] HashMap = new int[52];
        
        for(int i = 0 ; i<J.length() ; i++)
        {
            char ch = J.charAt(i);
  
            int hashIdx = hash(ch);          
            HashMap[hashIdx] = 1;    
        }
        
        int count = 0;
        
        for(int i = 0 ; i<S.length() ; i++)
        {
            char ch = S.charAt(i);

            int hashIdx = hash(ch);
            if( HashMap[hashIdx]==1) count++;
        }
        return count;   
    }
}


//Using inbuilt HashSet
// class Solution {
//     public int numJewelsInStones(String J, String S) 
//     {
//         HashSet<Character> hs = new HashSet<>();
//         for(int i = 0 ; i<J.length() ; i++)
//         {
//             char ch = J.charAt(i);
//             hs.add(ch);
//         }
//         int count = 0;
//         for(int i = 0 ; i<S.length() ; i++)
//         {
//             char ch = S.charAt(i);
//             if(hs.contains(ch)) count++;
//         }
//         return count;
//     }
// }